### ISSUE TITLE
Global variables in multi-streamed queries and mutations

### ISSUE THREAD

**Issue by lveillard** [vaticle/typedb#6778]
This would be a way to solve lots of non workaroundable issues and performance drops, both in queries and mutations. 

# The problem
## Queries
- When querying multiple entities, we are forced to split it into different streams, which have to be serialized. For instance if we are querying authors and their books, and we want to gather all the attributes linked to both, we need to do a first stream with authors, their attributes and the iids of books related to them. And then a second roundtrip for the books. This causes huge performance drops vs Neo4J or Dgraph.
- The result is obtained as a permutation of entities, relations and attributes. As an example, an outcome that appears as 12 nodes with their attributes in the graph structure in Studio, appears as 1440 lines in the query! (I will add the example as a comment)

## Mutations
- The main problem is that every match line has to be ok in order to modify the database, this means that we can't do "optional" inserts. As an example if we want to add 10 authors and add a certain book to evert existing author called "Robert" in one single stream, if there is no "Robert" author all the mutation is discarded. So we need two streams. The problem with needing two streams is that we lose the possibility to use the variables of the entitis we are creating in the stream 1 on the stream 2, with forces us to do a query after the first mutation stream, and then a mutation, with performance consequences

# Proposed solution
Being able to send in parallel, in one single roundtrip and one single collect() multiple streams where variables are global, and get the results in one mixed stream

## rules
- A transaction should be able to be sent, collected and reversed globally in one single roundtrip, including all its streams
- Variables must be global in order to gather the advantages of mixing different streams
- Independent constraints: Each stream depends on its match clauses but not on the other streams
   -  Exception to this are variables used in other streams that are null in the stream they are defined. 

## example in client

The client would like some thing like this pseudo code:
```
const transactionOne = new transaction;
transactionOne.add(stream1, stream2, stream3)
const results = await transacitonOne.collect();
```

I will add examples as comments

📝An obvious alternative would be to be able to create nested match-delete-insert queries in one single stream. The same rules would apply

**Comment by lveillard** [comment]
# Example 2: Example mutation

It is imposible to mix these two use cases:

## problem
Use case a) Generating an entity A which is related to entities B and C. 
- This is easy when done in one single query because we can reuse the $var of A, B, C in our mutation and make everything happen at once.

Use case b) Conditionally adding A to B if B is related to D
- This is possible adding "B is related to D" as a condition in the match part, and then relating A to B. The relation will only happen if B is related to D

use case a) & b) Doing both things at once seems to not possible.

Imagine that you want to to relate A to C, and also to B if B is related to D. As the constraint of B<->D will be a dependency for the full mutation, the full  A<->C might not happen. So you will need to split the mutation in two. A->C always, and A->B depending if B<->D. 
But if you split the mutation in two, you can't use the $VarA id on both streams. 

So it's either case a) or case b) but it seems impossible to do both at once.

## Solution

Each stream being dependent of their own match clauses + variables being able to be used globally, we could send in one round trip to the database like this:

```
stream
  insert
    $a isa author, has id 1;
    $b isa book, has id 2;
    $ab (book: $b, author: $a) isa authorshiop;
    
 stream
   match
     ($r, $otherthing) isa somerelation; #we cna imagine any other contraint here
   
   insert
      $r isa reviewer, has id 3;
      $ab (book: $b, reviewer: $a) isa review;
```

## Result:
Stream 1 will always work
Stream 2 is conditioned to the match of the stream 2, but does not prevent stream 1 from happening
We ere able to use var $a in both streams, so no need to query the database and do 3 serialized queries
One single round trip!

**Comment by lveillard** [comment]
# Example 3 Unlinking two connections on a nested relation

This is another non-workaroundable issue (optional would help, but this would also solve it)

imagine that you want to unlink two things of a single element. For simplicity, i'm showing a graphQL-like mutation

#get author 3 and unlink every book and link it to city London
```json
"$thing": 'Author',
"$id": '3',
"books": null , #replace to null
"city": '0x111' (london) //replace to null
"tags": {$arrayOp: 'add', value: "sci-fi'} //add to current
```
📝Notes
=> books = Null means **if** there is a relation to books, remove all them
=> City = London means, "if" there is a link to any other city, replace it (is a cardinality = ONE field)
=> ArrayOp add, respects current tags, so unlike the other two lines, is not a replace but a pure insert

There is no way to do this in one single stream in TypeDB.  With multristreams with global variables it would look like this

```bash
stream
  match 
   $a isa author, has id '3';

  insert
  $a has tag, 'sci-fi';

stream 
  match
  $r(book: $b, author: $a) isa authorship;
  
  delete 
  $r isa authorship;

stream
  match
   $r(city: $c, author: $a) isa livingIn;
   
   delete
  $r isa livingIn;

stream 
  match
  $c isa city, has name 'London';
  
  insert
  (author: $a, city: $london) isa livingIn

```

## Result
Like this, we are able to send in one single roundtrip to DB all the items we wanted, where some conditions are local to the insert/delete clauses they should affect and not the full transaction.

**Comment by lveillard** [comment]
# Example 4 Unlinking two optional roles in the same "hyper-relation"

So this is really similar to the example 3 and its also impossible to do atm. 

Imagine a relation like this:

```
insert
$recruitingProcess1 (candidate: $c, interview: $i1, $interview: $i2, job:$j)  isa process, has id '1';
$recruitingProcess2 (candidate: $c, job:$j)  isa process, has id '2';
```

Now imagine we want to do an operation like this:

```jsonc
{
 $thing: 'process',
 $id: [1,2],
 interviews: null,
 job: null
}
```
(get process with id 1 and 2 and unlink them to interviews or jobs if needed)

This would not be possible in one single stream. As matching both things would actually discard 'id2' who has no interviews

## Solution

```
stream
  match 
  $r ($a) isa process, has id in [1,2];

stream 
 match 
   $a isa interview

 delete 
  $r( interview: $a);
  
stream
  match
  $a isa job

  delete
  $r ( job: $r);
```


📝 Note 1: This example gets worse if multiple roles are played by the same type of entity
📝 Note 2:  I'm using 'in' as a sugarcoat of a more complex query

## Result
As a result we are able to remove interviews and jobs in both processes, even if id = 2 has no interviews already, the rest of the streams do work.

**Comment by lveillard** [comment]
# Example 5:  A really complex example

If i could simplify one single use case, would be something like this. The day I will be able to translate this to one single typeQL query, all our pain is solved:

```jsonc
[
  {
    "$thing": "User",
    "$op": "create",
    "name": "Peter",
    "email": "Peter@test.ru",
    "accounts": [
      { "$op": "create", "provider": "google" },
      { "$op": "create", "$tempId": "acc1", "provider": "facebook" }
    ]
  },
  {
    "$thing": "User",
    "op": "update",
    "email": "newEmail@gmail.com",
    "accounts": [{"$op": "link", "filter":[{"token": {"$eq": "bz434djuX0Eyd"}}]}],
  },
  {
    "$thing": "Accounts-Old",
    "$op": "delete"  //delete all Accounts-old
  },
  {
    "$thing": "User",
    "filter": {"email": "someuser@gmail.com"},
    "accounts": {"$op": "delete"} //delete evert acccount linked to this user
  }
  {
    "$tempId": "us1",
    "$op": "create",
    "$thing": "User",
    "name": "Bob"
  },
  {
    "$tempId": "gh1",
    "$op": "create",
    "$thing": "Account",
    "provider": "github"
  },
  {
    "$thing": "User",
    "$op": "create",
    "name": "Bea",
    "accounts": [
      { "$op": "create","provider": "facebook" },
      { "$op": "create", "$tempId": "acc1" },
      { "$op": "link", "$tempId": "gh1" },
      { "$op": "link", "$filter": { "$op": "create", "provider": "google" } }
    ]
  },
  {
    "$thing": "Account",
    "$op": "create",
    "provider": "Microsoft",
    "user": { "$op": "create", "name": "Carla" }
  }
]
```
## problem 
right now it is impossible to make it in one single transaction. And really complex to do in serialized streams.

## Solution
Same as in the other examples. Building multiple streams

## Results
- This would generate all the operations (create, update, delete,  unlink, link) in one single transaction. Filters would be added to the match clause, of each stream. Impacting only those streams, while variables would be able to be used globally.

- TempIds are important, because we are relating things in one single query (is equivalent to $gh1 isa account). Without multiple streams with shared variables, we should send tempID stuff first, and then do a second round trip doing matches to those created things. 

📝 words starting with $ are reserved words like $op, $thing, $id, $tempId, $eq... all the other words are just attributes or roles's names

**Comment by flyingsilverfin** [comment]
I think although the problems you point are valid:
1. being able to retrieve say sets of attributes attached to objects as part of the query
2. optionals in matches and inserts

will be addressed with other approaches. The first will be addressed soon (we will undertake some research on the best way to solve this in TypeDB) -- this quarter for sure. The second will likely wait until TypeDB 3.0 is completed, but may come sooner.

The last issue of reducing the permutativity of the answers doesn't have a clear solution yet, but we will also be investigating that while considering problem 1. above.

Thank you for all the effort and the proposals, but the approach outlined is not line with the direction we intend to move, so I will close the issue.
